### PR TITLE
[zh-cn]: update the translation of "Using WebRTC Encoded Transforms"

### PR DESCRIPTION
**Pull request by tianyeeT**
### Related issues and pull requests
#16473

### CODE REVIEW

File: files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+  });
+  received_video.srcObject = event.streams[0];
+};
+```
+
+再次注意，你可以在任何时候添加转换流。但是通过在 `track` 事件处理程序中添加它，可以确保转换流将获得轨道的第一帧编码帧。
+
+## Worker 实现
+
+工作线程脚本必须实现一个处理 {{domxref("DedicatedWorkerGlobalScope/rtctransform_event", "rtctransform")}} 事件的处理程序，创建一个 [管道链](/zh-CN/docs/Web/API/Streams_API/Concepts#pipe_chains)，将 `event.transformer.readable`（{{DOMxRef("ReadableStream")}}）流通过 {{DOMxRef("TransformStream")}} 传输到 `event.transformer.writable`（{{DOMxRef("WritableStream")}}）流中。
```
Comment (T34-active) [inline files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md:88]:
```suggestion
工作线程脚本必须实现一个处理 {{domxref("DedicatedWorkerGlobalScope/rtctransform_event", "rtctransform")}} 事件的处理程序，创建一个[链式管道](/zh-CN/docs/Web/API/Streams_API/Concepts#链式管道传输)，将 `event.transformer.readable`（{{DOMxRef("ReadableStream")}}）流通过 {{DOMxRef("TransformStream")}} 传输到 `event.transformer.writable`（{{DOMxRef("WritableStream")}}）流中。
```

File: files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+请注意，创建管道链的代码与上一个示例中的代码相同。
+
+### 运行时与变换进行通信
+
+[`RTCRtpScriptTransform` 构造函数](/zh-CN/docs/Web/API/RTCRtpScriptTransform/RTCRtpScriptTransform) 允许你传递选项和对象到工作线程。在前面的示例中，我们传递了静态信息，但有时你可能希望在运行时修改工作线程中的变换算法，或者从工作线程中获取信息。例如，支持加密的 WebRTC 会议可能需要向变换使用的算法添加一个新的密钥。
+
+虽然可以使用 {{domxref("Worker.postMessage()")}} 在运行变换代码的工作线程和主线程之间共享信息，但通常更容易使用 {{domxref("MessageChannel")}} 作为 [`RTCRtpScriptTransform` 构造函数](/zh-CN/docs/Web/API/RTCRtpScriptTransform/RTCRtpScriptTransform) 的选项共享频道，因为在处理新的编码帧时，通道上下文直接可在 `event.transformer.options` 中使用。
+
+以下代码创建了一个 {{domxref("MessageChannel")}} 并将其第二个端口 [传输](/zh-CN/docs/Web/API/Web_Workers_API/Transferable_objects) 给工作线程。主线程和变换随后可以使用第一个和第二个端口进行通信。
```
Comment (T34-active) [inline files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md:165]:
```suggestion
以下代码创建了一个 {{domxref("MessageChannel")}} 并将其第二个端口[传输](/zh-CN/docs/Web/API/Web_Workers_API/Transferable_objects)给工作线程。主线程和变换随后可以使用第一个和第二个端口进行通信。
```

File: files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+// 将 channel.port2 作为构造函数选项传递给 transform，并将其传输到 worker。
+const channel = new MessageChannel();
+const transform = new RTCRtpScriptTransform(
+  worker,
+  { purpose: "encrypt", port: channel.port2 },
+  [channel.port2],
+);
+
+// 使用 port1 发送一个字符串。（我们可以发送和传输基本类型/对象）。
+channel.port1.postMessage("A message for the worker");
```
Comment (T34-active) [inline files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md:181]:
```suggestion
channel.port1.postMessage("给 worker 的消息");
```

File: files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+请注意，你可以在任何时候添加转换，但是通过在调用 `addTrack()` 后立即添加转换，转换将获得发送的第一帧编码帧。
+
+## 添加用于传入帧的转换
+
+运行在 worker 中的转换通过将相应的 `RTCRtpScriptTransform` 分配给传入轨道的 {{domxref("RTCRtpReceiver.transform")}} 来插入到传入的 WebRTC 管道中。
+
+这个例子展示了如何添加一个转换来修改传入流。该代码假定已经连接到远程对等端的名为 `peerConnection` 的 {{domxref("RTCPeerConnection")}}。
+
+首先，我们添加了一个 `RTCPeerConnection` 的 [`track` 事件](/zh-CN/docs/Web/API/RTCPeerConnection/track_event) 处理程序，以捕获当对等端开始接收新轨道时的事件。在处理程序内部，我们构造了一个 `RTCRtpScriptTransform` 并将其添加到 `event.receiver.transform`（`event.receiver` 是一个 {{domxref("RTCRtpReceiver")}}）。
```
Comment (T34-active) [inline files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md:71]:
```suggestion
首先，我们添加了一个 `RTCPeerConnection` 的 [`track` 事件](/zh-CN/docs/Web/API/RTCPeerConnection/track_event)处理程序，以捕获当对等端开始接收新轨道时的事件。在处理程序内部，我们构造了一个 `RTCRtpScriptTransform` 并将其添加到 `event.receiver.transform`（`event.receiver` 是一个 {{domxref("RTCRtpReceiver")}}）。
```

File: files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md [hunk 5]
```diff
@@ -0,0 +1,260 @@
+---
+title: Using WebRTC Encoded Transforms
```
Comment (T34-active) [inline files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md:2]:
```suggestion
title: 使用 WebRTC 编码转换
```

File: files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+      controller.enqueue(encodedFrame);
+    },
+  });
+  event.transformer.readable
+    .pipeThrough(transform)
+    .pipeTo(event.transformer.writable);
+});
+```
+
+WebRTC 编码转换的实现类似于“通用” `TransformStream`，但存在一些重要的区别。像通用流一样，它的[构造函数](/zh-CN/docs/Web/API/TransformStream/TransformStream#parameters)接受一个定义了可选的 [`start()`](/zh-CN/docs/Web/API/TransformStream/TransformStream#startcontroller) 方法的对象，该方法在构造时调用，[`flush()`](/zh-CN/docs/Web/API/TransformStream/TransformStream#flushcontroller) 方法，在流即将关闭时调用，以及 [`transform()`](/zh-CN/docs/Web/API/TransformStream/TransformStream#transformchunk_controller) 方法，每当有一个块需要处理时都会调用。
```
Comment (T34-active) [inline files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md:126]:
```suggestion
WebRTC 编码转换的实现类似于“通用” `TransformStream`，但存在一些重要的区别。像通用流一样，它的[构造函数](/zh-CN/docs/Web/API/TransformStream/TransformStream#参数)接受一个定义了可选的 [`start()`](/zh-CN/docs/Web/API/TransformStream/TransformStream#startcontroller) 方法的对象，该方法在构造时调用，[`flush()`](/zh-CN/docs/Web/API/TransformStream/TransformStream#flushcontroller) 方法，在流即将关闭时调用，以及 [`transform()`](/zh-CN/docs/Web/API/TransformStream/TransformStream#transformchunk_controller) 方法，每当有一个块需要处理时都会调用。
```

**Review by T34-active — COMMENTED** [review]
fix content

File: files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+  const { rid, key } = event.data;
+  // 密钥由转换器用于加密帧（未显示）
+
+  // 使用 rid 获取编解码器生成新关键帧。这里的 'rcevent' 是 rtctransform 事件。
+  rcevent.transformer.generateKeyFrame(rid);
+};
+```
+
+接收方请求新关键帧的代码几乎相同，只是没有指定“rid”。
+这里是仅包含端口消息处理程序的代码：
```
Comment (skyclouds2001) [inline files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md:235]:
```suggestion
接收方请求新关键帧的代码几乎相同，只是没有指定“rid”。这里是仅包含端口消息处理程序的代码：
```

File: files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+
+### 触发关键帧
+
+原始视频很少被发送或存储，因为用完整图像来表示每一帧会消耗大量的空间和带宽。相反，编解码器定期生成一个包含足够信息构建完整图像的“关键帧”，在关键帧之间发送“增量帧”，它们只包含自上一个增量帧以来的变化。虽然这比发送原始视频要高效得多，但这意味着为了显示与特定增量帧相关联的图像，你需要最后一个关键帧和所有随后的增量帧。
+
+这可能会导致新用户加入 WebRTC 会议应用时出现延迟，因为他们在收到第一个关键帧之前无法显示视频。同样，如果使用编码转换来加密帧，则接收方在收到使用其密钥加密的第一个关键帧之前无法显示视频。
+
+为了确保在需要时尽早发送新的关键帧，`event.transformer` 中的 {{domxref("RTCRtpScriptTransformer")}} 对象有两种方法：{{domxref("RTCRtpScriptTransformer.generateKeyFrame()")}}，它会导致编解码器生成一个关键帧，和 {{domxref("RTCRtpScriptTransformer.sendKeyFrameRequest()")}}，它会导致接收方可以从发送方请求一个关键帧。
+
+下面的示例显示了主线程如何将加密密钥传递给发送方转换，并触发编解码器生成一个关键帧。请注意，主线程无法直接访问 {{domxref("RTCRtpScriptTransformer")}} 对象，因此它需要将密钥和限制标识符（"rid" 是流 ID，指示必须生成关键帧的编码器）传递给 worker。在这里，我们使用了一个 `MessageChannel`，使用了与前一节相同的模式。代码假定已经有一个对等连接，并且 `videoSender` 是一个 {{domxref("RTCRtpSender")}}。
```
Comment (skyclouds2001) [inline files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md:202]:
```suggestion
下面的示例显示了主线程如何将加密密钥传递给发送方转换，并触发编解码器生成一个关键帧。请注意，主线程无法直接访问 {{domxref("RTCRtpScriptTransformer")}} 对象，因此它需要将密钥和限制标识符（“rid”是流 ID，指示必须生成关键帧的编码器）传递给 worker。在这里，我们使用了一个 `MessageChannel`，使用了与前一节相同的模式。代码假定已经有一个对等连接，并且 `videoSender` 是一个 {{domxref("RTCRtpSender")}}。
```

File: files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+// 使用 port1 发送一个字符串。（我们可以发送和传输基本类型/对象）。
+channel.port1.postMessage("给 worker 的消息");
+channel.port1.start();
+```
+
+在 worker 中，端口可作为 `event.transformer.options.port` 使用。下面的代码显示了如何监听端口的 `message` 事件以从主线程获取消息。你还可以使用该端口将消息发送回主线程。
+
+```js
+event.transformer.options.port.onmessage = (event) => {
+  // 消息载荷在 'event.data' 中
```
Comment (skyclouds2001) [inline files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md:189]:
```suggestion
  // 消息载荷在“event.data”中
```

File: files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+// 创建一个 channel
+// 将 channel.port2 作为构造函数选项传递给 transform，并将其传输到 worker。
+const channel = new MessageChannel();
+const transform = new RTCRtpScriptTransform(
+  worker,
+  { purpose: "encrypt", port: channel.port2 },
+  [channel.port2],
+);
+
+// 使用 port1 发送一个字符串。（我们可以发送和传输基本类型/对象）。
```
Comment (skyclouds2001) [inline files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md:180]:
```suggestion
// 使用 port1 发送一个字符串。（我们可以发送和传输基本类型/对象）
```

File: files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+    .pipeTo(event.transformer.writable);
+});
+```
+
+WebRTC 编码转换的实现类似于“通用” `TransformStream`，但存在一些重要的区别。像通用流一样，它的[构造函数](/zh-CN/docs/Web/API/TransformStream/TransformStream#参数)接受一个定义了可选的 [`start()`](/zh-CN/docs/Web/API/TransformStream/TransformStream#startcontroller) 方法的对象，该方法在构造时调用，[`flush()`](/zh-CN/docs/Web/API/TransformStream/TransformStream#flushcontroller) 方法，在流即将关闭时调用，以及 [`transform()`](/zh-CN/docs/Web/API/TransformStream/TransformStream#transformchunk_controller) 方法，每当有一个块需要处理时都会调用。
+
+与通用构造函数不同，任何在构造函数对象中传递的 `writableStrategy` 或 `readableStrategy` 属性都会被忽略，队列策略完全由用户代理管理。
+
+`transform()` 方法也不同，它接收的是 {{DOMxRef("RTCEncodedVideoFrame")}} 或 {{DOMxRef("RTCEncodedAudioFrame")}}，而不是通用的“块”。
+除了它展示了如何将帧转换为可以修改并在之后排队到流上的形式之外，此处显示的方法并不引人注目。
```
Comment (skyclouds2001) [inline files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md:131]:
```suggestion
`transform()` 方法也不同，它接收的是 {{DOMxRef("RTCEncodedVideoFrame")}} 或 {{DOMxRef("RTCEncodedAudioFrame")}}，而不是通用的“块”。除了它展示了如何将帧转换为可以修改并在之后排队到流上的形式之外，此处显示的方法并不引人注目。
```

File: files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+
+### 使用单独的发送器和接收器转换
+
+之前的例子在发送和接收时使用相同的转换函数时可以工作，但在许多情况下，算法会有所不同。你可以为发送器和接收器使用单独的 worker 脚本，或者在一个 worker 中处理这两种情况，如下所示。
+
+如果 worker 用于发送器和接收器，它需要知道当前的编码帧是来自编解码器的传出帧，还是来自封包器的传入帧。可以使用 [`RTCRtpScriptTransform` 构造函数](/zh-CN/docs/Web/API/RTCRtpScriptTransform/RTCRtpScriptTransform)的第二个选项来指定此信息。
+例如，我们可以为发送器和接收器定义一个单独的 `RTCRtpScriptTransform`，传递相同的 worker 和一个 options 对象，其中的 `name` 属性指示转换是用于发送还是接收（如上面的前几节所示）。然后在 worker 中，可以通过 `event.transformer.options` 获取到此信息。
```
Comment (skyclouds2001) [inline files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md:138]:
```suggestion
如果 worker 用于发送器和接收器，它需要知道当前的编码帧是来自编解码器的传出帧，还是来自封包器的传入帧。可以使用 [`RTCRtpScriptTransform` 构造函数](/zh-CN/docs/Web/API/RTCRtpScriptTransform/RTCRtpScriptTransform)的第二个选项来指定此信息。例如，我们可以为发送器和接收器定义一个单独的 `RTCRtpScriptTransform`，传递相同的 worker 和一个 options 对象，其中的 `name` 属性指示转换是用于发送还是接收（如上面的前几节所示）。然后在 worker 中，可以通过 `event.transformer.options` 获取到此信息。
```

File: files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+
+### 触发关键帧
+
+原始视频很少被发送或存储，因为用完整图像来表示每一帧会消耗大量的空间和带宽。相反，编解码器定期生成一个包含足够信息构建完整图像的“关键帧”，在关键帧之间发送“增量帧”，它们只包含自上一个增量帧以来的变化。虽然这比发送原始视频要高效得多，但这意味着为了显示与特定增量帧相关联的图像，你需要最后一个关键帧和所有随后的增量帧。
+
+这可能会导致新用户加入 WebRTC 会议应用时出现延迟，因为他们在收到第一个关键帧之前无法显示视频。同样，如果使用编码转换来加密帧，则接收方在收到使用其密钥加密的第一个关键帧之前无法显示视频。
+
+为了确保在需要时尽早发送新的关键帧，`event.transformer` 中的 {{domxref("RTCRtpScriptTransformer")}} 对象有两种方法：{{domxref("RTCRtpScriptTransformer.generateKeyFrame()")}}，它会导致编解码器生成一个关键帧，和 {{domxref("RTCRtpScriptTransformer.sendKeyFrameRequest()")}}，它会导致接收方可以从发送方请求一个关键帧。
+
+下面的示例显示了主线程如何将加密密钥传递给发送方转换，并触发编解码器生成一个关键帧。请注意，主线程无法直接访问 {{domxref("RTCRtpScriptTransformer")}} 对象，因此它需要将密钥和限制标识符（“rid”是流 ID，指示必须生成关键帧的编码器）传递给 worker。在这里，我们使用了一个 `MessageChannel`，使用了与前一节相同的模式。代码假定已经有一个对等连接，并且 `videoSender` 是一个 {{domxref("RTCRtpSender")}}。
```
Comment (JasonLamv-t) [inline files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md:200]:
```suggestion
下面的示例显示了主线程如何将加密密钥传递给发送方转换，并触发编解码器生成一个关键帧。请注意，主线程无法直接访问 {{domxref("RTCRtpScriptTransformer")}} 对象，因此它需要将密钥和限制标识符（“rid”是流 ID，指示必须生成关键帧的编码器）传递给工作线程。在这里，我们使用了一个 `MessageChannel`，使用了与前一节相同的模式。代码假定已经有一个对等连接，并且 `videoSender` 是一个 {{domxref("RTCRtpSender")}}。
```

File: files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+
+// 将 rid 和新密钥发送到发送方
+channel.port1.start();
+channel.port1.postMessage({
+  rid: "1",
+  key: "93ae0927a4f8e527f1gce6d10bc6ab6c",
+});
+```
+
+在 worker 中的 {{domxref("DedicatedWorkerGlobalScope/rtctransform_event", "rtctransform")}} 事件处理程序获取端口，并使用它来监听来自主线程的 `message` 事件。如果收到事件，则获取 `rid` 和 `key`，然后调用 `generateKeyFrame()`。
```
Comment (JasonLamv-t) [inline files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md:220]:
```suggestion
在工作线程中的 {{domxref("DedicatedWorkerGlobalScope/rtctransform_event", "rtctransform")}} 事件处理程序获取端口，并使用它来监听来自主线程的 `message` 事件。如果收到事件，则获取 `rid` 和 `key`，然后调用 `generateKeyFrame()`。
```

**Review by JasonLamv-t — CHANGES_REQUESTED** [review]

File: files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+
+`transform()` 方法也不同，它接收的是 {{DOMxRef("RTCEncodedVideoFrame")}} 或 {{DOMxRef("RTCEncodedAudioFrame")}}，而不是通用的“块”。除了它展示了如何将帧转换为可以修改并在之后排队到流上的形式之外，此处显示的方法并不引人注目。
+
+### 使用单独的发送器和接收器转换
+
+之前的例子在发送和接收时使用相同的转换函数时可以工作，但在许多情况下，算法会有所不同。你可以为发送器和接收器使用单独的 worker 脚本，或者在一个 worker 中处理这两种情况，如下所示。
+
+如果 worker 用于发送器和接收器，它需要知道当前的编码帧是来自编解码器的传出帧，还是来自封包器的传入帧。可以使用 [`RTCRtpScriptTransform` 构造函数](/zh-CN/docs/Web/API/RTCRtpScriptTransform/RTCRtpScriptTransform)的第二个选项来指定此信息。例如，我们可以为发送器和接收器定义一个单独的 `RTCRtpScriptTransform`，传递相同的 worker 和一个 options 对象，其中的 `name` 属性指示转换是用于发送还是接收（如上面的前几节所示）。然后在 worker 中，可以通过 `event.transformer.options` 获取到此信息。
+
+在这个例子中，我们在全局专用 worker 范围对象上实现了 `onrtctransform` 事件处理程序。`name` 属性的值用于确定构造哪个 `TransformStream`（实际的构造方法没有显示）。
```
[the remark this answers is not in the document]
Comment (yin1999) [reply files/zh-cn/web/api/webrtc_api/using_encoded_transforms/index.md]: - worker 不翻译
- “scope”应被翻译为“作用域”
- “event handler”应被翻译为“事件处理器”

**Review by JasonLamv-t — APPROVED** [review]
LGTM. Thank you! @tianyeeT 👍🏼

**Pull request merged**